spatial/r3/transform.go

```diff
@@ -0,0 +1,267 @@
+	}
+	t = t.Translate(Scale(-1, origin))
+	t = t.scale(factor)
+	return t.Translate(origin)
+}
+
+func (t Transform) scale(factor Vec) Transform {
+	t.d00 = (t.d00+1)*factor.X - 1
+	t.x10 *= factor.X
+	t.x20 *= factor.X
+	t.x30 *= factor.X
+
+	t.x01 *= factor.Y
+	t.d11 = (t.d11+1)*factor.Y - 1
+	t.x21 *= factor.Y
+	t.x31 *= factor.Y
+
+	t.x02 *= factor.Z
+	t.x12 *= factor.Z
+	t.d22 = (t.d22+1)*factor.Z - 1
+	t.x32 *= factor.Z
+	return t
+}
+
+// Mul multiplies the Transforms a and b and returns the result.
+// This is the equivalent of combining two transforms in one.
+func (t Transform) Mul(b Transform) Transform {
+	if t == (Transform{}) {
+		return b
+	}
+	if b == (Transform{}) {
+		return t
+	}
+	x00 := t.d00 + 1
+	x11 := t.d11 + 1
+	x22 := t.d22 + 1
+	x33 := t.d33 + 1
+	y00 := b.d00 + 1
+	y11 := b.d11 + 1
+	y22 := b.d22 + 1
+	y33 := b.d33 + 1
+	var m Transform
+	m.d00 = x00*y00 + t.x01*b.x10 + t.x02*b.x20 + t.x03*b.x30 - 1
+	m.x10 = t.x10*y00 + x11*b.x10 + t.x12*b.x20 + t.x13*b.x30
+	m.x20 = t.x20*y00 + t.x21*b.x10 + x22*b.x20 + t.x23*b.x30
+	m.x30 = t.x30*y00 + t.x31*b.x10 + t.x32*b.x20 + x33*b.x30
+	m.x01 = x00*b.x01 + t.x01*y11 + t.x02*b.x21 + t.x03*b.x31
+	m.d11 = t.x10*b.x01 + x11*y11 + t.x12*b.x21 + t.x13*b.x31 - 1
+	m.x21 = t.x20*b.x01 + t.x21*y11 + x22*b.x21 + t.x23*b.x31
+	m.x31 = t.x30*b.x01 + t.x31*y11 + t.x32*b.x21 + x33*b.x31
+	m.x02 = x00*b.x02 + t.x01*b.x12 + t.x02*y22 + t.x03*b.x32
+	m.x12 = t.x10*b.x02 + x11*b.x12 + t.x12*y22 + t.x13*b.x32
+	m.d22 = t.x20*b.x02 + t.x21*b.x12 + x22*y22 + t.x23*b.x32 - 1
+	m.x32 = t.x30*b.x02 + t.x31*b.x12 + t.x32*y22 + x33*b.x32
+	m.x03 = x00*b.x03 + t.x01*b.x13 + t.x02*b.x23 + t.x03*y33
+	m.x13 = t.x10*b.x03 + x11*b.x13 + t.x12*b.x23 + t.x13*y33
+	m.x23 = t.x20*b.x03 + t.x21*b.x13 + x22*b.x23 + t.x23*y33
+	m.d33 = t.x30*b.x03 + t.x31*b.x13 + t.x32*b.x23 + x33*y33 - 1
+	return m
+}
+
+// Det returns the determinant of the Transform.
+func (t Transform) Det() float64 {
+	x00 := t.d00 + 1
+	x11 := t.d11 + 1
+	x22 := t.d22 + 1
+	x33 := t.d33 + 1
+	return x00*x11*x22*x33 - x00*x11*t.x23*t.x32 +
+		x00*t.x12*t.x23*t.x31 - x00*t.x12*t.x21*x33 +
+		x00*t.x13*t.x21*t.x32 - x00*t.x13*x22*t.x31 -
+		t.x01*t.x12*t.x23*t.x30 + t.x01*t.x12*t.x20*x33 -
+		t.x01*t.x13*t.x20*t.x32 + t.x01*t.x13*x22*t.x30 -
+		t.x01*t.x10*x22*x33 + t.x01*t.x10*t.x23*t.x32 +
+		t.x02*t.x13*t.x20*t.x31 - t.x02*t.x13*t.x21*t.x30 +
+		t.x02*t.x10*t.x21*x33 - t.x02*t.x10*t.x23*t.x31 +
+		t.x02*x11*t.x23*t.x30 - t.x02*x11*t.x20*x33 -
+		t.x03*t.x10*t.x21*t.x32 + t.x03*t.x10*x22*t.x31 -
+		t.x03*x11*x22*t.x30 + t.x03*x11*t.x20*t.x32 -
+		t.x03*t.x12*t.x20*t.x31 + t.x03*t.x12*t.x21*t.x30
+}
+
+// Inv returns the inverse of the transform such that
+// t.Inv() * t is the identity Transform.
+// If matrix is singular then Inv() returns the zero transform.
+func (t Transform) Inv() Transform {
+	if t == (Transform{}) {
+		return t
+	}
+	det := t.Det()
+	if math.Abs(det) < 1e-16 {
+		return zeroTransform
+	}
+	// Do something if singular?
```

Comment:
This troubles me. In mat we've gone out of our way to make it possible to avoid needing to calculate the inverse of matrices.

Reply:
I'm not sure how I'd go about avoiding this calculation without using the `mat` package which would make the implementation noticeably slower. I'll add the benchmark source in this next push
```
goos: linux
goarch: amd64
pkg: gonum.org/v1/gonum/spatial/r3
cpu: AMD Ryzen 5 3400G with Radeon Vega Graphics    
BenchmarkTransformInverse-8      	12232880	        93.51 ns/op	       0 B/op	       0 allocs/op
BenchmarkMatTransformInverse-8   	  438561	      2564 ns/op	      96 B/op	       4 allocs/op
PASS
ok  	gonum.org/v1/gonum/spatial/r3	5.583s
```